### PR TITLE
Gradle upgrade

### PR DESCRIPTION
**Pull request by ddv239**
Upgrade Gradle, add annotations support and .editorconfig.

### CODE REVIEW

File: gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,7 +1,7 @@
...
-zipStoreBase=GRADLE_USER_HOME
-zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-3.3-all.zip
-distributionSha256Sum=71a787faed83c4ef21e8464cc8452b941b5fcd575043aa29d39d15d879be89f7
+#Fri Jan 26 17:08:07 PST 2018
+distributionBase=GRADLE_USER_HOME
+distributionPath=wrapper/dists
+zipStoreBase=GRADLE_USER_HOME
+zipStorePath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.1-all.zip
```
Comment (Adonai) [inline gradle/wrapper/gradle-wrapper.properties]: Hmmm, the README mentions Gradle 4.5 but wrapper downloads 4.1

Comment (ddv239) [reply gradle/wrapper/gradle-wrapper.properties]: I am not sure how various components are versioned. I upgraded the Gradle "plugin" in the current version of Android Studio and it gave me this version. When I downloaded the current version of Graddle itself, to verify that I did not break the command-line build, it was 4.5, and I had do use a non-latest version of the JDK to make it work, so I added a note on that to readme in case someone else hits the same problem. I see no contradiction here.

File: .editorconfig [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+# top-most EditorConfig file
+root=true
+
+[*]
+charset=utf-8
+# Unix-style newlines with a newline ending every file
+end_of_line=lf
+insert_final_newline=true
+indent_style=tab
+tab_width=4
```
Comment (adrian-bl) [inline .editorconfig]: This is not related to gradle, please revert this change.

File: README.md [hunk 3]
```diff
@@ -55,15 +55,17 @@ Building
...
 Building from command-line #1
 --------------------------
+> Note: at the time of this writing, the current version of Gradle ([4.5](https://gradle.org/releases/)) is not compatible with the current version of JDK ([9.0.4](http://www.oracle.com/technetwork/java/javase/downloads/jdk9-downloads-3848520.html)). To have the build succeed, use JDK version [1.8.0_162](http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html).
  * `gradle build` to build the APK
  * Optional: `gradle installDebug` to install the APK to a connected device
  
 Building from command-line #2
 --------------------------
+> Note: at the time of this writing, the current version fo the Android SDK tools are not compatible with ant, see https://stackoverflow.com/questions/42912824/the-ant-folder-is-suddenly-missing-from-android-sdk-did-google-remove-it
  * `android update project --path .` to generate local.properties
```
Comment (adrian-bl) [inline README.md]: We can remove this section as ant is not supported anymore.

File: README.md [hunk 4]
```diff
@@ -55,15 +55,17 @@ Building
 To build you will need:
 
  * A Java compiler compatible with Java 1.7
```
Comment (adrian-bl) [inline README.md]: s/1.7/1.8/

File: gradle/wrapper/gradle-wrapper.properties [hunk 5]
```diff
@@ -1,7 +1,7 @@
...
+distributionSha256Sum=71a787faed83c4ef21e8464cc8452b941b5fcd575043aa29d39d15d879be89f7
```
Comment (adrian-bl) [inline gradle/wrapper/gradle-wrapper.properties]: This sha256 sum is wrong.

File: gradle/wrapper/gradle-wrapper.properties (hunk 1 above)
Comment (adrian-bl) [reply gradle/wrapper/gradle-wrapper.properties]: Version mappings are listed here:
https://developer.android.com/studio/releases/gradle-plugin.html#updating-gradle

File: build.gradle [hunk 6]
```diff
@@ -52,3 +53,7 @@ android {
         abortOnError false
     }
 }
+
+dependencies {
+    compileOnly  'com.android.support:support-annotations:27.0.2'
```
Comment (adrian-bl) [inline build.gradle]: I still don't get what exactly we get from this.
Is it an Android-Studio thing?

File: gradle/wrapper/gradle-wrapper.properties (hunk 1 above)
Comment (ddv239) [reply gradle/wrapper/gradle-wrapper.properties]: OK, I can say that with this change, everything works perfectly correctly. Gradle 4.1 serves this change's purpose. The comment in Readme was about the "current version of Gradle" which is never than 4.1.

File: build.gradle (hunk 6 above)
Comment (ddv239) [reply build.gradle]: As mentioned before in a comment to #716: it's just annotations part of it for making code look cleaner (e.g. @NonNull or @Nullable etc. JSR-305). It is not specific to Android Studio and is used during compilation.

File: .editorconfig (hunk 2 above)
Comment (ddv239) [reply .editorconfig]: OK, will make a separate PR

File: README.md (hunk 4 above)
Comment (ddv239) [reply README.md]: I don't understand what this comment means

File: .editorconfig (hunk 2 above)
Comment (adrian-bl) [reply .editorconfig]: I don't think we need to clutter the repo with this.
Every sane editor detects line-endings and tab/whitespace automatically anyway.

File: README.md (hunk 4 above)
Comment (adrian-bl) [reply README.md]: Replace 1.7 with 1.8, as the text below (correctly) states that java >= 1.8 is required.

File: .editorconfig (hunk 2 above)
Comment (ddv239) [reply .editorconfig]: I understand you but I have a sane Android Studio editor, and even though it automatically detects coding conventions already present in the sources and keeps my changes consistent with them, it pops a yellow warning bar on top of the text editor telling me that it replaced spaces with tabs to follow the document's formatting style but that does not match my editor preferences.
.editorconfig is the clean way to avoid any such inconsistencies and tell any sane editor what are the project's preferences that should override any globally configured user's preferences.
I truly believe you should not resist this clean and standard way to achieve this consistency and you have to admit that this "clutter" is truly minimal and so not worth avoiding.
By the way, presence of .editorconfig would even make it unnecessary to mention the corresponding formatting conventions in the Readme (which it currently does).

File: README.md (hunk 3 above)
Comment (ddv239) [reply README.md]: Sure

File: README.md (hunk 4 above)
Comment (ddv239) [reply README.md]: Actually, you already replaced that in a previous change so my updated version of this PR will not have this line changed.

File: gradle/wrapper/gradle-wrapper.properties (hunk 5 above)
Comment (ddv239) [reply gradle/wrapper/gradle-wrapper.properties]: Good catch; I just had android Studio make this change for me and did not realize it failed to properly update the checksum. Will make sure the updated version does contain a valid one.

**Comment by ddv239** [comment]
I have pushed a new state to this branch which isolates the "Gradle upgrade" only, with no annotations support or .editorconfig additions, addressing all the comments.
